gadget/update.go

```diff
@@ -321,38 +331,52 @@ func RemodelUpdatePolicy(from, to *LaidOutStructure) (bool, *LaidOutStructure, *
 //
 // But any non-kernel assets need to be ignored, they will be handled by
 // the regular gadget->gadget update mechanism and policy.
-func KernelUpdatePolicy(from, to *LaidOutStructure) (bool, *LaidOutStructure, *LaidOutStructure) {
-	var kernelContent []ResolvedContent
+func KernelUpdatePolicy(from, to *LaidOutStructure) (bool, ResolvedContentFilterFunc) {
 	for _, rn := range to.ResolvedContent {
 		if rn.KernelUpdateFlag {
-			kernelContent = append(kernelContent, rn)
+			return true, func(rn *ResolvedContent) bool {
+				return rn.KernelUpdateFlag
+			}
 		}
 	}
-	if len(kernelContent) > 0 {
-		newTo := *to
-		newTo.ResolvedContent = kernelContent
-		return true, from, &newTo
-	}
 
-	return false, nil, nil
+	return false, nil
 }
 
 func resolveUpdate(oldVol *PartiallyLaidOutVolume, newVol *LaidOutVolume, policy UpdatePolicyFunc) (updates []updatePair, err error) {
 	if len(oldVol.LaidOutStructure) != len(newVol.LaidOutStructure) {
 		return nil, errors.New("internal error: the number of structures in new and old volume definitions is different")
 	}
-	for j := range oldVol.LaidOutStructure {
+	for j, oldStruct := range oldVol.LaidOutStructure {
+		newStruct := newVol.LaidOutStructure[j]
 		// update only when new edition is higher than the old one; boot
 		// assets are assumed to be backwards compatible, once deployed
 		// are not rolled back or replaced unless a higher edition is
 		// available
-		if needsUpdate, from, to := policy(&oldVol.LaidOutStructure[j], &newVol.LaidOutStructure[j]); needsUpdate {
-			updates = append(updates, updatePair{from, to})
+		if update, filter := policy(&oldStruct, &newStruct); update {
+			updates = append(updates, updatePair{
+				from: &oldVol.LaidOutStructure[j],
+				to:   &newVol.LaidOutStructure[j],
+
+				resolvedContentFilter: filter,
+			})
 		}
 	}
 	return updates, nil
 }
 
+func filterUpdates(updates []updatePair) {
+	for _, update := range updates {
+		if update.resolvedContentFilter != nil {
+			for i := len(update.to.ResolvedContent) - 1; i >= 0; i-- {
+				if !update.resolvedContentFilter(&update.to.ResolvedContent[i]) {
+					update.to.ResolvedContent = append(update.to.ResolvedContent[:i], update.to.ResolvedContent[i+1:]...)
```

Comment:
this is not the most readable/natural way of doing the filtering though,  making a fresh list seems better, I also don't know if we need to copy update.to first overall (I don't think is strictly necessary but might be more robust)